### PR TITLE
feat(recipes): add GPUDirect-RDMA torch-distributed-rdma runtime for GB200 GKE Kubeflow

### LINKED ISSUE #2837: [Feature]: Add a GPUDirect-RDMA-wired torch-distributed-rdma ClusterTrainingRuntime for the GB200 GKE Kubeflow leaf (parity with H100/TCPXO)
### Prerequisites

- [x] I searched existing issues

### Feature Summary

The `gb200-gke-cos-training-kubeflow` recipe leaf (GB200/A4X on GKE COS) ships
only the generic `torch-distributed` Kubeflow `ClusterTrainingRuntime`, which
carries none of the GPUDirect-RDMA fabric wiring the cluster needs. The sibling
H100 leaf (`h100-gke-cos-training-kubeflow`) does not have this gap: it ships
both the generic `torch-distributed` runtime AND a fabric-wired
`torch-distributed-tcpxo` runtime
(`recipes/components/kubeflow-trainer/manifests/torch-distributed-tcpxo-cluster-training-runtime.yaml`).
GB200 needs the equivalent — a `torch-distributed-rdma` `ClusterTrainingRuntime`
pre-wired for GPUDirect-RDMA over the `gke-gb200-rdma` component's fixed-name
Network objects (`gvnic-1`, `rdma-0`..`rdma-3`).

### Problem/Use Case

Today, a Kubeflow user submitting a TrainJob against the GB200/GKE recipe gets
the plain `torch-distributed` runtime by default. That runtime requests GPUs
per the submitted TrainJob (not a fixed whole node), sets no RDMA network
attachments, mounts no gIB host libraries, and sets no NCCL_NET=gIB — so it
either silently falls back to the slower default network path or fails,
despite the cluster having a fully qualified GPUDirect-RDMA fabric
provisioned and validated by `gke-gb200-rdma`'s own health check.

This gap was raised during review on PR #2338: the integrator docs describe
recipes' NCCL workloads setting NCCL_NET=gIB and using all four RDMA NICs, but
the Kubeflow leaf selects the generic torch-distributed runtime, which
supplies neither the six-network annotation, gIB host mounts, nor NCCL
environment — so a workload launched through the bundled runtime can use the
ordinary network path or fail instead of getting the documented fail-closed
GPUDirect-RDMA behavior.

That PR resolved the immediate review finding by narrowing the docs to scope
the fail-closed guarantee to the validator only, and documenting what a
custom TrainJob must add by hand for the same guarantee — a correct interim
fix, but it leaves the actual capability gap open: there is still no way for
a Kubeflow user to get a working, wired GPUDirect-RDMA runtime out of the box
on GB200/GKE, unlike on H100/GKE (TCPXO) where one already exists.

### Proposed Solution

Add `recipes/components/kubeflow-trainer/manifests/torch-distributed-rdma-cluster-training-runtime.yaml`,
a `ClusterTrainingRuntime` named `torch-distributed-rdma`, modeled on
`torch-distributed-tcpxo-cluster-training-runtime.yaml` but wired for GB200's
fixed RDMA topology instead of H100's recipe-recorded interface mapping
(no --gke-tcpxo-interfaces-style flag should be needed here, since
gvnic-1/rdma-0..3 are the same fixed names on every GB200/GKE cluster).
Reference implementation for the required wiring already exists in
`validators/performance/testdata/gb200/gke/runtime-nvls.yaml`'s "node" job:

- Pod annotations:
  `networking.gke.io/default-interface: eth0` and
  `networking.gke.io/interfaces` mapping eth1→gvnic-1, eth2..eth5→rdma-0..3
- hostPath volumes for `/home/kubernetes/bin/nvidia` and `/home/kubernetes/bin/gib`
  (read-only), mounted into the container
- `NCCL_NET=gIB` (and sourcing `set_nccl_env.sh` if present, matching the
  validator's convention)
- `IPC_LOCK` capability
- Whole-node GPU request: `nvidia.com/gpu: "4"` (limits and requests), matching
  the a4x-highgpu-4g node shape and the RDMA-can't-be-shared-between-pods
  constraint already documented in gke-gb200-networking.md
- mlPolicy.torch (not MPI) at numNodes: 2 minimum, consistent with
  torch-distributed-tcpxo's rationale that a single-node default for a
  multi-node fabric runtime would exercise nothing

Then wire it into the leaf, mirroring the H100 pattern exactly:

  recipes/overlays/gb200-gke-cos-training-kubeflow.yaml:
```
    componentRefs:
      - name: kubeflow-trainer
        manifestFiles:
          - .../torch-distributed-cluster-training-runtime.yaml
          - .../torch-distributed-rdma-cluster-training-runtime.yaml   # new
```

Also update docs/integrator/gke-gb200-networking.md to mention the new
pre-wired runtime once it exists (rather than only describing what a custom
TrainJob must add manually), and update docs/user/component-catalog.md /
docs/user/recipe-health.md if the new manifest changes any published counts.

### Success Criteria

- `torch-distributed-rdma-cluster-training-runtime.yaml` exists and renders
  with all 5 fixed-name network attachments, gIB host mounts, NCCL_NET=gIB,
  IPC_LOCK, and a whole-node (4x) GPU request — no unset/optional fabric
  wiring paths (unlike TCPXO, there's no recipe-recorded value that could be
  missing, so this should never render "visibly broken").
- `gb200-gke-cos-training-kubeflow.yaml` ships both the generic and the RDMA
  runtime, matching the H100 leaf's `manifestFiles` shape.
- `pkg/bundler/testdata/stock_render_golden.yaml`,
  `pkg/recipe/testdata/catalog_parity_golden.yaml`, and
  `pkg/recipe/testdata/coverage_golden.yaml` regenerated and green via
  `make update-goldens` / `TestCatalogParityGolden` /
  `TestStockRenderParityGolden`.
- New/updated tests: a render test analogous to the TCPXO coverage in
  `recipes/gke_gb200_rdma_test.go` or `pkg/recipe/metadata_test.go`
  (`TestGB200GKEIncludesRDMA`-style) asserting the new runtime is present for
  this leaf.
- `go run ./tools/evidence-pointercheck` and `go vet ./...` clean.
- docs/integrator/gke-gb200-networking.md no longer needs to tell a custom
  TrainJob author to hand-roll the wiring — it can point at the shipped
  runtime instead (or state both options).

### Alternatives Considered

Leave the current doc-only resolution (PR #2338) as the permanent answer and
never ship a wired runtime for GB200 Kubeflow. Rejected: this leaves a real
capability/parity gap vs. the H100 leaf and pushes the fabric wiring burden
onto every downstream Kubeflow user by hand, with no working reference other
than the validator's own internal test fixture.

### Component

Recipe engine / data

### Priority

Important (would improve my workflow)

### Compatibility / Breaking Changes

None — additive manifest + additive manifestFiles entry; no existing behavior changes.

### Operational Considerations

None beyond the existing GB200/GKE RDMA prerequisites already documented in gke-gb200-networking.md (multi-networking CRs, nccl-rdma-installer DaemonSet — both already required by this leaf today).

### Are you willing to contribute?

Yes, I can open a PR

### PR DESCRIPTION
**Pull request by mikecook**
## Summary

Adds `torch-distributed-rdma`, a GPUDirect-RDMA-wired Kubeflow `ClusterTrainingRuntime`
for the `gb200-gke-cos-training-kubeflow` leaf, so a TrainJob gets the fabric wiring by
name instead of hand-rolling it.

## Motivation / Context

The GB200 GKE Kubeflow leaf shipped only the generic `torch-distributed` runtime, which
carries no RDMA network attachments, no gIB host mounts, and no `NCCL_NET=gIB`. A job
submitted against it falls back to TCP or fails, despite the cluster having a qualified
GPUDirect-RDMA fabric provisioned and validated by the `gke-gb200-rdma` component.

The H100 leaf already closes this gap with `torch-distributed-tcpxo`. PR #2338 resolved
the immediate review finding by narrowing the docs, which was correct but left the
capability gap open.

Fixes: #2837
Related: #2338

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Build/CI/tooling

## Component(s) Affected

- [ ] CLI (`cmd/aicr`, `pkg/cli`)
- [ ] API server (`cmd/aicrd`, `pkg/server`)
- [x] Recipe engine / data (`pkg/recipe`)
- [ ] Bundlers (`pkg/bundler`, `pkg/component/*`)
- [ ] Collectors / snapshotter (`pkg/collector`, `pkg/snapshotter`)
- [ ] Validator (`pkg/validator`)
- [ ] Core libraries (`pkg/errors`, `pkg/k8s`)
- [x] Docs/examples (`docs/`, `examples/`)
- [ ] Other: ____________

## Implementation Notes

The runtime carries the `networking.gke.io` annotations mapping `eth1` to `gvnic-1` and
`eth2` through `eth5` to `rdma-0` through `rdma-3`, the `/home/kubernetes/bin/nvidia` and
`/home/kubernetes/bin/gib` host mounts, `NCCL_NET=gIB`, `IPC_LOCK`, and a whole-node
4-GPU request. The wiring matches the NVLS validator fixture
(`validators/performance/testdata/gb200/gke/runtime-nvls.yaml`) mount for mount.

**No recipe-generation input.** `gvnic-1` and `rdma-0` through `rdma-3` are fixed object
names on every GB200/GKE cluster, so the manifest renders unconditionally and has no
unset path that could ship it half-wired. This is why no `--gke-tcpxo-interfaces`
equivalent is introduced.

**Deliberate deviation from the issue.** The issue asked for `NCCL_NET=gIB` "and sourcing
`set_nccl_env.sh` if present, matching the validator's convention." The runtime does not
source it. Kubeflow Trainer owns the container command, so a `ClusterTrainingRuntime` has
no shell hook to source from. A workload that wants upstream's tuning profile sources it
from its own `spec.trainer.command`, which is documented with a working example. Skipping
it costs tuning, not correctness, because `NCCL_NET=gIB` is what selects the fabric.

Structural parity with `torch-distributed-tcpxo` is exact on every shared element (Helm
hook annotations, the `framework: torch` and `trainjob-ancestor-step` labels, and both
`acceleratedNodeSelector` / `acceleratedTolerations` injection blocks). The only
difference is TCPXO's `tcpxo-daemon` sidecar, which RDMA does not need because gIB
installs on the host via DaemonSet.

Docs also now warn that setting `resourcesPerNode` on a TrainJob replaces the runtime's
resources outright on Trainer v2.2.0, dropping the `nvidia.com/gpu: "4"` request. That is
worse here than on TCPXO, because RDMA cannot be shared between Pods on a node.

## Testing

```bash
# Commands run (prefer `make qualify` for non-trivial changes)
make qualify
```

`make qualify` passes. The two gates the issue names explicitly are clean:
`go run ./tools/evidence-pointercheck` and `go vet ./...`.

Goldens regenerated and verified current by re-running `make update-goldens` to an empty
diff: `stock_render_golden.yaml` and `catalog_parity_golden.yaml` changed,
`coverage_golden.yaml` correctly needed no change. `make recipe-health-docs` produces no
change for this leaf, because the R/D/P/C columns count declared validation checks rather
than manifests. No BOM update, because the runtime reuses the `pytorch/pytorch` digest
already listed under `kubeflow-trainer`.

Both new tests were mutation-verified rather than assumed:

- Render test (`recipes/gke_gb200_rdma_test.go`): flipping `NCCL_NET` to `socket`,
  typo'ing either scheduling `with` key, and drifting one network name in the interfaces
  annotation each fail it.
- Resolver test (`pkg/recipe/gb200_gke_kubeflow_rdma_test.go`): deleting the manifest
  line from the overlay fails it, while the render test still passes, which is the
  coverage split that justifies having both.

No coverage delta to report. The only Go files added are tests, so there is no new
exported surface and no uncovered production code.

## Risk Assessment

- [x] **Low** — Isolated change, well-tested, easy to revert
- [ ] **Medium** — Touches multiple components or has broader impact
- [ ] **High** — Breaking change, affects critical paths, or complex rollout

Purely additive. One new manifest plus one new `manifestFiles` entry on a single leaf.
No existing behavior changes.

**Rollout notes:** None. The generic `torch-distributed` runtime still ships on the same
leaf and is unchanged, so existing TrainJobs keep resolving exactly as before. Opting in
means setting `spec.runtimeRef.name: torch-distributed-rdma`. The cluster prerequisites
(multi-networking CRs, `nccl-rdma-installer`) were already required by this leaf.

## Checklist

- [x] Tests pass locally (`make test` with `-race`)
- [x] Linter passes (`make lint`)
- [x] I did not skip/disable tests to make CI green
- [x] I added/updated tests for new functionality
- [x] I updated docs if user-facing behavior changed
- [x] Changes follow existing patterns in the codebase
- [x] Commits are cryptographically signed (`git commit -S`) — [GPG signing info](https://docs.github.com/en/authentication/managing-commit-signature-verification)

### CODE REVIEW

File: docs/integrator/gke-gb200-networking.md [hunk 1]
```diff
@@ -142,6 +137,64 @@ output; that one is GKE-managed (created automatically once
...
+```yaml
+spec:
+  trainer:
+    command: ["/bin/bash", "-c"]
+    args:
+      - |
+        if [ -f /home/kubernetes/bin/gib/scripts/set_nccl_env.sh ]; then
+          . /home/kubernetes/bin/gib/scripts/set_nccl_env.sh
+        fi
+        exec torchrun --nnodes "$PET_NNODES" --nproc-per-node 4 train.py
```
[the remark this answers is not in the document]
Comment (mikecook) [reply docs/integrator/gke-gb200-networking.md]: Addressed in 61298bab0, though not quite as suggested, because the stated failure mode does not occur.

`torchrun` reads those values itself. In `torch/distributed/run.py` at v2.11.0, the version pinned by this runtime's `pytorch/pytorch:2.11.0-cuda12.8-cudnn9-runtime` image, all three flags are declared with `action=env`, which resolves `PET_{DEST}` before falling back to the default.

```
--node-rank      PET_NODE_RANK     default=0
--master-addr    PET_MASTER_ADDR   default="127.0.0.1"
--master-port    PET_MASTER_PORT   default=29500
```

`--rdzv-backend` defaults to `static`, and the endpoint is then derived from those two values.

```python
def get_rdzv_endpoint(args):
    if args.rdzv_backend == "static" and not args.rdzv_endpoint:
        return f"{args.master_addr}:{args.master_port}"
    return args.rdzv_endpoint
```

Trainer v2.2.0 sets all five variables, which your own research established from `constants.go`. So the documented two-node job does form a single group with a bare `torchrun`, and the proposed flags restate the environment the container already carries.

The comment did catch a real defect, just not the one it named. The example passed `--nnodes` and `--nproc-per-node` while omitting the other three, which reads as a complete flag set and is not one. Hardcoding `--nproc-per-node 4` is worse than redundant, since Trainer derives `PET_NPROC_PER_NODE` from the GPU limit, making it the one flag here that can contradict the job it runs under.

The fix is to pass none of them rather than all of them, which is also the end state upstream describes in the Trainer rendezvous issue your research surfaced, where the goal is `torchrun train.py`. The example is now:

```yaml
        exec torchrun train.py
```

with a sentence explaining that Trainer sets the `PET_*` variables and `torchrun` reads each as the default for the matching flag.

**Review by mchmarny — COMMENTED** [review]
Comment: no findings against 61298ba. GitHub reports a merge conflict; required checks and focused recipe tests otherwise pass at the reviewed SHA.

**Comment by mikecook** [comment]
@coderabbit resume

**Review by mchmarny — APPROVED** [review]
Approve: no new findings against b58c1b0. Focused recipe tests pass and GitHub now reports the head mergeable.
